Make User activity list look like Project activity list

cbliard opened the pull request:
https://community.openproject.org/work_packages/45344

There is code and logic duplication as there are two ways of displaying an activity list since work package 45083 has been done.

The aim of this PR is to use the same rendering in both places, and remove dead code used for the previous rendering.

cbliard committed:
Render project activity list with ViewComponent

cbliard committed:
Render user activity list like in project activity

Rename 'op-project-activity' prefix in css classes as 'op-activity' as
this css class is used in multiple contexts.

cbliard committed:
refactor: Remove unused parameters

cbliard committed:
Do not display user in user activity list

And internationalize strings in activity list templates

ulferts commented on app/components/activity_item_component.rb:
```diff
@@ -0,0 +1,65 @@
+           :format_time,
+           :link_to_user,
+           :truncate_formatted_text,
+           to: :helpers
```
I believe this to be a left-over. As far as I can see, the methods are always send directly to `helpers` and some of those listed here seem to by now have moved into another component.

cbliard replied:
You are right, this is a left-over. Thanks.

ulferts commented on app/components/activity_item_component.rb:
```diff
@@ -0,0 +1,65 @@
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class ActivityItemComponent < ViewComponent::Base
```
Currently all the components are at `app/components`. That works for the current situation where there are but few of them. To ease adding more components later on, I suggest to start with placing the activity components into their own directory (e.g. `app/components/activities`).

cbliard replied:
Done

ulferts commented on app/components/activity_item_component.html.erb:
```diff
@@ -0,0 +1,25 @@
+    <% end -%>
+  </div>
+  <%=
+    render(ActivityItemSubtitleComponent.new(user: @display_user && @event.event_author,
```
I had to look up the call stack to understand why `@display_user && @event.event_author` was used here as I initially thought it was a typo and `||` would be used instead. It would have been easier to read for me if it would be something like `display_user? && @event.event_author` were used instead. It just wasn't clear to me that `display_user` was a boolean and not an actual user object.

cbliard replied:
Good point. Will change it.

ulferts commented on app/components/activity_item_subtitle_component.rb:
```diff
@@ -0,0 +1,59 @@
+  end
+
+  def i18n_key
+    i18n_key = 'activity.item.'.dup
```
Why have a `.dup` here?

cbliard replied:
To make the string mutable.
I try adding `# frozen_string_literal: true` for all new files I create, and sometimes for some files I modify. Strings modified explicitly need to be unfrozen with `+s` or `s.dup`. As I find `+s` cryptic I use `s.dup`. I agree that it does not really reveal the intent of having a mutable string.

[Documentation](https://ruby-doc.org/core-3.1.1/String.html#method-i-2B) about it is terse.

ulferts commented on app/components/activity_list_component.html.erb:
```diff
@@ -0,0 +1,6 @@
+  <% @events.each do |event| -%>
+    <% journal = @journals_by_id[event.event_id] -%>
+    <%= render(ActivityItemComponent.new(event:, journal:, display_user: @display_user)) %>
+  <% end -%>
```
Nitpicking as I am admittedly picky right now since I want to understand the patterns to use. Those 4 lines are all pure ruby code, no tag at all. Since there is an underlying ruby class now, that is one of the benefits of `ViewComponent` I would have expected to have as much non template code in the `component.rb` class as possible. 

So I would at least have preferred something like: 

```
<% journals.each do |journal| %>
   <%= render(ActivityItemComponent.new(event:, journal:, display_user: @display_user)) %>
<% end %>
```

Where `journals` would be the collection of all journals associated to the `@events`. That would also ease the transition to only working with journals as well.

But I am wondering whether I would even prefer:

```
   <%= render_journals %>
```

since there are no tags. Today, I would prefer it for the reason that I find a mix of languages harder to read. Tomorrow, it might be different, though.

cbliard replied:
Not sure if I addressed this one with my last commit. 

Now it's
```
  <% @events.each do |event| -%>
    <%= render(Activities::ItemComponent.new(event:, display_user: @display_user)) %>
  <% end -%>
```

You would like to have it put into the `.rb` file?

cbliard replied:
I found another way to do it:

```
<%= render(Activities::ItemComponent.with_collection(@events, display_user: @display_user)) %>
```

[It handles the iteration + rendering part of it](https://viewcomponent.org/guide/collections.html).

ulferts commented on app/controllers/users_controller.rb:
```diff
@@ -85,11 +84,24 @@ def show
+      .index_by(&:id)
+
+    render layout: 'no_menu'
+  end
```
This is almost identical to the code in the `activities_controller`. Please extract this into a shared method.

cbliard replied:
Yes, almost identical. I was not sure about how to extract it. A shared method is not the right abstraction IMHO. event and journal should be tied together at a higher level.

I'll try to refactor it in a better way to remove this logic duplication.

cbliard replied:
Should be ok now with latest commit.

ulferts commented on app/controllers/users_controller.rb:
```diff
@@ -85,11 +84,24 @@ def show
-        format.html { render layout: 'no_menu' }
+        format.html do
+          show_respond_html(events)
+        end
```
We don't really need the `respond_to` block any more. There is only one entry in it so for the client, it should not make a difference at all. 
We could also add a file-wide `respond_to :html` to the top of the file to have it clarified for all methods.

cbliard replied:
Good suggestion. Fixed.

ulferts commented on app/views/users/show.html.erb:
```diff
@@ -80,24 +80,11 @@ See COPYRIGHT and LICENSE files for more details.
+        <% @events_by_day.sort_by { |day, _events| day }.reverse_each do |day, events| -%>
+          <h4 class="op-activity-days--header"><%= format_activity_day(day) %></h3>
+
+          <%= render(ActivityListComponent.new(events:, journals_by_id: @journals_by_id, display_user: false)) %>
```
I am probably getting greedy but we could move the 'activity by day' functionality into its own view component. The `id="activity"` can probably be remove alltogether so the only difference between this code here and the one in the `activities/index.html` seems to be the `display_user: false` flag.
Isn't a must though, just an opportunity.

cbliard replied:
There is also the header which is different: `h4` in user activity list, `h3` in the project activity list.

I'll try to see what I can do to use it in only one place.

cbliard replied:
Should be ok now with latest commit.

ulferts commented on frontend/src/global_styles/content/_list.sass:
```diff
@@ -1,67 +0,0 @@
-// See COPYRIGHT and LICENSE files for more details.
-//++
-
-.generic-list
```
You seem to be right that this is not used any more. Well spotted, then 👍

ulferts commented on spec/features/users/show_spec.rb:
```diff
@@ -0,0 +1,51 @@
+
+  let(:index_page) { Pages::Admin::Users::Index.new }
+
+  it 'works almost correctly', with_settings: { journal_aggregation_time_minutes: 0 } do
```
Hmm, what is that supposed to tell me? I am scared now 😆 😱

cbliard replied:
Thanks for spotting it. It is a placeholder text from a snippet I frequently use.
Fixed.

ulferts reviewed (changes requested):
A good change overall. I only have some small things and questions.

cbliard committed:
refactor: clearer code

cbliard committed:
refactor: only respond to html

cbliard committed:
refactor: Move activity components in a dedicated folder

cbliard committed:
refactor: Reduce code duplication

`Activities::DaysComponent` factorizes the code to generate the HTML for
the activities per days.

By moving `journal` into `Activities::Event`, `journals_by_id` does not
need to be passed around anymore. Eager loading of journals is moved to
`Activities::Fetcher` instead of being handled by the controller. It
also means that the atom feed will be slower as it does not need the
journals information.

cbliard committed:
refactor: use ViewComponent collection rendering to reduce view code size

ulferts commented on spec/features/users/show_spec.rb:
```diff
@@ -0,0 +1,51 @@
+    expected_work_package_title = "#{work_package.type.name} ##{work_package.id}: #{work_package.subject} " \
+                                  "(Project: #{work_package.project.name})"
+    expect(page).to have_selector('li', text: expected_work_package_title)
+  end
```
Good point about adding a feature spec for this previously untested part of the application.

ulferts commented on app/components/activities/item_component.rb:
```diff
@@ -0,0 +1,64 @@
+#++
+
+class Activities::ItemComponent < ViewComponent::Base
+  with_collection_parameter :event
```
That is a nice addition to avoid loops in the template.

cbliard replied:
Yep, I was happy to find this. Thanks for pushing for a better code.

ulferts reviewed (approved):
Definitely feels like an improvement over the structure we had before this PR. 

I learned quite a lot reviewing the PR, thanks.
